### ISSUE TITLE
extunix fails to link with `undefined reference to backtrace_symbols' on Alpine

### ISSUE THREAD

**Issue by talex5** [ygrek/extunix#42]
This seems to affect anything using extunix, including its own tests:
```
docker pull ocaml/opam:alpine-ocaml-4.12
docker run --rm -it ocaml/opam:alpine-ocaml-4.12
opam install -t extunix
# + /home/opam/.opam/4.12/bin/ocamlfind ocamlopt -g -I src/ -linkpkg -package bigarray -package bytes -package oUnit -package unix -I src -I test src/extunix.cmxa test/test.cmx -o test/test.native
# /usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: src//libextunix_stubs.a(execinfo.o): in function `caml_extunix_backtrace':
# /home/opam/.opam/4.12/.opam-switch/build/extunix.0.2.0/_build/src/execinfo.c:13: undefined reference to `backtrace'
# /usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: /home/opam/.opam/4.12/.opam-switch/build/extunix.0.2.0/_build/src/execinfo.c:15: undefined reference to `backtrace_symbols'
# collect2: error: ld returned 1 exit status
# File "caml_startup", line 1:
# Error: Error during linking (exit code 1)
# Command exited with code 2.
# E: Failure("Command ''/home/opam/.opam/4.12/bin/ocamlbuild' src/libextunix_stubs.a src/dllextunix_stubs.so src/extunix.cma src/extunix.cmxa src/extunix.a src/extunix.cmxs test/test.native test/testba.native test/test_user_namespace.native -tag debug -tag tests' terminated with error code 10")
```

This affects extunix 0.2.0. The current Git version seems to work, though, so maybe it just needs a new release?

(possibly related fix: https://github.com/ocaml-multicore/ocaml-multicore/pull/482)

**Comment by ygrek** [comment]
there is new release in opam PR now, but I am confused how can it work given there were no related changes :thinking:

**Comment by MisterDA** [comment]
If I'm not mistaken, this patch should fix it: https://github.com/ygrek/extunix/pull/43/commits/ebb4d377753c1711763f3ba9b3c348e58a4532cb.
The problem is indeed that `libexecinfo` isn't part of libc in Alpine. I've added a depext to the package, and the required compiler link flag (but I haven't tested it with the CI yet).

**Comment by talex5** [comment]
Hmm, but that commit isn't merged yet, right? And running my test above with the opam-repository PR works fine!
```
docker pull ocaml/opam:alpine-ocaml-4.12
docker run --rm -it ocaml/opam:alpine-ocaml-4.12
(cd opam-repository && git pull origin pull/18483/head && opam update)
opam install -t extunix
-> installed extunix.0.3.1
Done.
```

**Comment by MisterDA** [comment]
I think I've got it: the `discover` script (the configure step) did generate C snippet but did not check if the snippet *linked* correctly, only if it *compiled* successfully.
https://github.com/ygrek/extunix/blob/016744879e00cc5b33e76654a0e742728dfc0afa/src/discover.ml#L98-L99
That's the case with the Alpine image, as the `libexecinfo-dev` package is installed, and the header `execinfo.h` is found, with the two symbols (you can see that other functions that are not available under musl are correctly not found). But then the link step isn't done. Simply running the linker on the obj file triggers the error.
Dune-configurator however tries to build a linked executable. It correctly skips the execinfo module if the library isn't linked, and if we teach the discover script the link flag, it can pick the functions up.

So I'd say this bug was fixed in the process of moving to Dune and Dune configurator, and thus is fixed in version 0.3 and onward.

**Comment by talex5** [comment]
Ah, that explains it then - thanks for tracking it down!
